File: Algorithms/leetcode/Java/AlgorithmsStudy/src/main/java/com/example/PopulatingNextRightPointersInEachNode.java
package com.example;

public class PopulatingNextRightPointersInEachNode {
    public static class Node {
        public int val;
        public Node left;
        public Node right;
        public Node next;

        public Node(int _val, Node _left, Node _right, Node _next) {
            val = _val;
            left = _left;
            right = _right;
            next = _next;
        }

        public Node() {
            this(0);
        }

        public Node(int _val) {
            this(_val, null, null, null);
        }
    };

    // https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/
    public Node connect(Node node) {
        for (Node i = node; i != null;) {
            Node newHead = null;
            Node newTail = null;
            for (Node j = i; j != null; j = j.next) {
                if (j.left != null) {
                    if (newTail != null) {
                        newTail.next = j.left;
                    } else {
                        newHead = j.left;
                    }
                    newTail = j.left;
                }
                if (j.right != null) {
                    if (newTail != null) {
                        newTail.next = j.right;
                    } else {
                        newHead = j.right;
                    }
                    newTail = j.right;
                }
                if (newTail != null) {
                    newTail.next = null;
                }
            }
            i = newHead;
        }
        return node;
    }
}

File: Algorithms/leetcode/TypeScript/src/populating.next.right.pointers.in.each.node.ts
/*
You are given a perfect binary tree where all leaves are on the same level, and every parent has two children. The binary tree has the following definition:

struct Node {
    int val;
    Node *left;
    Node *right;
    Node *next;
}
Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.

Initially, all next pointers are set to NULL.


Follow up:
    - You may only use constant extra space.
    - Recursive approach is fine, you may assume implicit stack space does not count as extra space for this problem.
*/

export class Node {
    val: number;
    left: Node | null;
    right: Node | null;
    next: Node | null;

    constructor(val?: number, left?: Node, right?: Node, next?: Node) {
        this.val = (val===undefined ? 0 : val);
        this.left = (left===undefined ? null : left);
        this.right = (right===undefined ? null : right);
        this.next = (next===undefined ? null : next);
    }
}

export function connect(root: Node | null): Node | null {
    let retval = root;
    if (root !== null) {
        root.next = null;
        let parent = root;
        let child = null;
        let previous = null;
        while (parent !== null) {
            for (let p = parent; p !== null; p = p.next) {
                if (p.left !== null) {
                    if (previous !== null) {
                        previous.next = p.left;
                    } else {
                        child = p.left;
                    }
                    previous = p.left;
                }
                if (p.right !== null) {
                    if  (previous !== null) {
                        previous.next = p.right;
                    } else {
                        child = p.right;
                    }
                    previous = p.right;
                }
            }
            if (previous !== null) {
                previous.next = null;
            }
            parent = child;
            child = null;
            previous = null;
        }
    }
    return retval;
}
